Restart the entire upload on retry, fixes #387.

goto-bus-stop opened the pull request:
Instead of restarting only the file upload itself in the Tus or
XHRUpload plugin, start an entirely new upload for retries. This fixes
retrying uploads with processing plugins, like Transloadit. Previously,
only the Tus upload would be retried--with this patch, a new assembly
will be created and the retry will go to that assembly.

arturi commented:
👍 I thought I’ve moved all the retry-related things into core methods, but turns out not 🙈 

So this will be creating a new tus/xhr upload on each retry. Not problematic memory-wise with possibly too many abandoned upload instances? Not sure if it’s possible to clean them somehow. Should we call `abort` before abandoning?

> tus: clear uploadURL when retrying

a little confused on this one. why, don’t we want to pick that up with each retry?

Looks good to me, I only looked through the code though, not gonna do testing right now (it’s probably also hard to do a real test, since there should be a way to make transcoding service fail for that?)

goto-bus-stop commented:
> Not problematic memory-wise with possibly too many abandoned upload instances?

Yeaaaah … we do keep around the event listeners too which could be a problem when retrying, because then `core:retry-upload` and other events can fire for multiple `tus.Upload` instances for the same file…

Perhaps ideally we should add one listener for each event in the tus plugin, and track the tus.Upload instances. Then when an event comes in we can do `this.uploads[file.id].abort()`, or whatever.

I'll try removing the event listeners when we receive a `retry` event. I think that should address it well enough for now. This is a good catch, thanks 🙏 

>  Should we call abort before abandoning?

hmm my thinking was that the upload will have errored by the time we retry. but maybe it's best to do that anyway.

> > tus: clear uploadURL when retrying
>
> a little confused on this one. why, don’t we want to pick that up with each retry?

Yeah we do, I changed that back in the next commit. we want to start at 0 again when used with Transloadit, since a retry creates a new assembly and should upload to the new assembly. the next commit moved the `uploadUrl: null` to the Transloadit plugin.

ifedapoolarewaju commented on src/core/Core.js:
```diff
@@ -486,16 +483,36 @@ class Uppy {
 
   reset () {
     this.cancelAll()
     // this.pauseAll()
```
we might as well delete this one too 😄

goto-bus-stop replied:
haha yeah, fixed :D

goto-bus-stop commented:
Now when starting an upload for a file, the Tus plugin will first clean up any previous uploads it still has for that file. Uploads, events and UppySockets are stored on the plugin instance and removed when an upload starts, fails, or succeeds. I took a bit of a brute force approach and added calls to the cleanup function in a bunch of places … maybe it's not necessary everywhere. It's definitely fixed the issue where multiple tus Upload objects responded to a single event. I think that's what was causing the tus locking errors also, multiple tus Uploads would attempt to resume an upload to the same file and upload the same chunks.

I'll throw a bunch more files at it and release 0.20.3 in the morning if it seems ok.

arturi commented:
Very nice move with `EventTracker` and `resetUploaderReferences` 👏  Thank you!

Should we later add it to `XHRUpload` too for UppySocket and events?
